File: examples/cost_bomb.js
/**
 * JavaScript Cost Bomb Example
 * 
 * This file demonstrates patterns that FinLinter should detect.
 * These are common anti-patterns that cause excessive cloud costs.
 */

const axios = require('axios');

// Simulated database connection
const db = {
    collection: (name) => ({
        find: (query) => Promise.resolve([]),
        findOne: (query) => Promise.resolve({}),
    })
};

/**
 * BAD: Multiple API calls inside a loop.
 */
async function processOrders(orderIds) {
    const results = [];

    for (const orderId of orderIds) {
        // BAD: fetch() call in loop
        const response = await fetch(`https://api.example.com/orders/${orderId}`);
        const data = await response.json();

        // BAD: axios call in loop
        const enrichment = await axios.get(`https://api.example.com/enrichment/${orderId}`);

        // BAD: JSON.stringify in loop
        const serialized = JSON.stringify(data);

        results.push({ data, enrichment: enrichment.data, serialized });
    }

    return results;
}


/**
 * BAD: Database calls in forEach loop.
 */
async function enrichUsers(users) {
    const enriched = [];

    users.forEach(async (user) => {
        // BAD: MongoDB find inside forEach
        const profile = await db.collection('profiles').findOne({ userId: user.id });

        // BAD: API call inside forEach
        const external = await fetch(`https://api.external.com/user/${user.id}`);

        enriched.push({ ...user, profile });
    });

    return enriched;
}


/**
 * BAD: Promise creation inside loop causes unbounded fan-out.
 */
function fanOutRequests(items) {
    const promises = [];

    for (let i = 0; i < items.length; i++) {
        // BAD: Promise created in loop - unbounded concurrency
        const promise = new Promise((resolve) => {
            fetch(`https://api.example.com/items/${items[i]}`)
                .then(r => r.json())
                .then(resolve);
        });
        promises.push(promise);
    }

    return Promise.all(promises);
}


/**
 * BAD: Database aggregation in loop.
 */
async function handleRequest(req) {
    const categories = req.body.categories;

    for (const category of categories) {
        // BAD: MongoDB aggregate in loop
        const stats = await db.collection('products').aggregate([
            { $match: { category } },
            { $group: { _id: null, total: { $sum: 1 } } }
        ]);

        // BAD: JSON.parse in loop
        const parsed = JSON.parse(req.body.data);
    }
}


/**
 * BAD: Array method with API calls.
 */
async function mapWithFetch(ids) {
    // BAD: fetch inside map (even with Promise.all, each creates a request)
    const results = await Promise.all(
        ids.map(async (id) => {
            const res = await fetch(`https://api.example.com/${id}`);
            return res.json();
        })
    );

    return results;
}


// ============================================================
// GOOD PATTERNS (for comparison - should NOT trigger warnings)
// ============================================================

/**
 * GOOD: Single batch API call.
 */
async function processOrdersCorrectly(orderIds) {
    // GOOD: Single batch request
    const response = await fetch('https://api.example.com/orders/batch', {
        method: 'POST',
        body: JSON.stringify({ ids: orderIds }),
    });

    const allData = await response.json();

    // Processing without I/O is fine
    const results = orderIds.map(id => ({
        id,
        data: allData[id]
    }));

    // GOOD: Single serialization at the end
    return JSON.stringify(results);
}

File: examples/cost_bomb.py
"""
Python Cost Bomb Example

This file demonstrates patterns that FinLinter should detect.
These are common anti-patterns that cause excessive cloud costs.
"""

import json
import requests

# Simulated imports for demonstration
class dynamodb:
    @staticmethod
    def get_item(TableName, Key):
        pass

class session:
    @staticmethod
    def query(model):
        pass

class db:
    @staticmethod
    def get(key):
        pass


def process_users(user_ids: list):
    """
    BAD: Multiple database and API calls inside a loop.
    This will trigger FinLinter warnings.
    """
    results = []
    
    for user_id in user_ids:
        # BAD: Database call in loop
        user_data = dynamodb.get_item(
            TableName='users',
            Key={'id': user_id}
        )
        
        # BAD: HTTP request in loop
        response = requests.get(
            f'https://api.example.com/enrichment/{user_id}'
        )
        
        # BAD: JSON serialization in loop
        serialized = json.dumps(user_data)
        
        results.append({
            'user': user_data,
            'enrichment': response.json(),
            'serialized': serialized
        })
    
    return results


def handle_batch_request(items: list):
    """
    BAD: ORM query inside a list comprehension (hidden loop).
    """
    # BAD: Database query in list comprehension
    enriched = [session.query(item) for item in items]
    
    # BAD: API calls in generator expression
    responses = (requests.post('https://api.example.com/process', json=item) 
                 for item in items)
    
    return list(responses)


def controller_endpoint(request_data: dict):
    """
    BAD: This is in a hot code path (controller name).
    """
    ids = request_data.get('ids', [])
    
    for id in ids:
        # BAD: Database get in loop in hot path
        data = db.get(id)
        
        # BAD: JSON loads in loop
        parsed = json.loads(data)
    
    return {'status': 'done'}


# ============================================================
# GOOD PATTERNS (for comparison - should NOT trigger warnings)
# ============================================================

def process_users_correctly(user_ids: list):
    """
    GOOD: Batch operations outside the loop.
    """
    # GOOD: Single batch call
    all_users = dynamodb.batch_get_item(user_ids)
    
    # GOOD: Single API call for batch
    enrichment = requests.post(
        'https://api.example.com/batch-enrichment',
        json={'user_ids': user_ids}
    )
    
    results = []
    for user_id in user_ids:
        # Processing without I/O is fine
        user = all_users.get(user_id)
        results.append(user)
    
    # GOOD: Single serialization at the end
    return json.dumps(results)
